### PR TITLE
Boundary input - Periodic

### PR DESCRIPTION
**Pull request by RevathiJambunathan**
This PR is a first in a series to follow to enhance boundary condition.
In this PR, a new user-defined input for field and particle boundary is introduced.
The idea is to support  (https://github.com/ECP-WarpX/WarpX/projects/14)
1. pec
2. pmc 
3. periodic
4. pml
The implementation will override `geometry.is_periodic`. 
Until all the above boundary condition support is added, if the user has specified 'geometry.is_periodic', then the `boundary.field_lo` and `boudnary.field_hi` will not be parsed.
Once, we have the pec,  pmc, pml all supported through the new `boundary.` interface, I will change the input scripts for all the CI-tests and expose the Abort added to the code, to ensure `is_periodic` is not used.

```
##########################################
###### Boundary Condition ################
##########################################
boundary.field_lo =periodic periodic periodic
boundary.field_hi =periodic periodic periodic
boundary.particle_lo = periodic periodic periodic
boundary.particle_hi = periodic periodic periodic
```

I have tested the PR for periodic boundaries for fields and particles

**Field periodicity test:
[test_field_periodicity.txt](https://github.com/ECP-WarpX/WarpX/files/6025010/test_field_periodicity.txt)

periodic_BC_Ey

**particle periodicity test
[inputs_test_particleAndField_periodicity.txt](https://github.com/ECP-WarpX/WarpX/files/6025017/inputs_test_particleAndField_periodicity.txt)
particles_periodic

### CODE REVIEW

**Comment by dpgrote** [comment]
Hi @RevathiJambunathan We have a collision here. I just submitted PR #1728 which is adding new capability for the particle boundary conditions, adding reflecting and allowing different boundaries on each side. Would you be able to have a chat to discuss the two PRs and combine our efforts? The main differences I have are a different set of input parameters and I added a new class to handle the boundary conditions.

**Comment by RevathiJambunathan** [comment]
> Hi @RevathiJambunathan We have a collision here. I just submitted PR1728 which is adding new capability for the particle boundary conditions, adding reflecting and allowing different boundaries on each side. Would you be able to have a chat to discuss the two PRs and combine our efforts? The main differences I have are a different set of input parameters and I added a new class to handle the boundary conditions.

Hi @dpgrote . Thank you for adding this capability!!!!
I was going to implement PEC for the fields next, and it looks like #1728 has done this for the particles!

Yes, I can meet to discuss this.
The project here describes what we had in mind
https://github.com/ECP-WarpX/WarpX/projects/14

**Comment by RevathiJambunathan** [comment]
Thank you for the offline discussion about merging the particle and field boundary structs @dpgrote !
I have separated the particle and field boundary types in the commit below
https://github.com/ECP-WarpX/WarpX/pull/1730/commits/77e70c05c709866ba09122a44268747e8548b75d

I have not moved it yet to the Source/BoundaryCondition, since during our discussion, we thought of renaming it as Source/FieldBoundaryCondition because all particle boundaries are in the particle container. 

So I was not sure if the boundary selection struct would go there?
Please let me know if you have any suggestions. Thank you

File: Source/Utils/WarpXAlgorithmSelection.H [hunk 1]
```diff
@@ -91,7 +91,35 @@ struct LoadBalanceCostsUpdateAlgo {
...
+struct FieldBoundaryType {
+    enum {
+        PEC = 0,     //!< perfect electric conductor (PEC) with E_tangential=0
+        Periodic = 1,
+        PML = 2,
+        PMC = 3      //!< perfect magnetic conductor (PMC) with B_tangential=0
+    };
+};
+
+/** Field boundary conditions at the domain boundary
```
Comment (RevathiJambunathan) [inline Source/Utils/WarpXAlgorithmSelection.H]:
```suggestion
/** Particle boundary conditions at the domain boundary
```

**Review by dpgrote — COMMENTED** [review]
Thanks for the changes! Can you add documentation for the new boundary input parameters? Or is that planned for a later PR?

File: Source/Utils/WarpXUtil.cpp [hunk 2]
```diff
@@ -349,6 +349,71 @@ void CheckGriddingForRZSpectral ()
...
+            if (particle_boundary_specified) {
+                AMREX_ALWAYS_ASSERT_WITH_MESSAGE(
+                    (WarpX::particle_boundary_lo[idim] == ParticleBoundaryType::Periodic) &&
+                    (WarpX::particle_boundary_hi[idim] == ParticleBoundaryType::Periodic),
+               "field and particle boundary must be periodic in both lo and hi");
+            } else {
+                // set particle boundary to periodic
+                WarpX::particle_boundary_lo[idim] = ParticleBoundaryType::Periodic;
+                WarpX::particle_boundary_hi[idim] = ParticleBoundaryType::Periodic;
+                amrex::Warning("Particle boundary is set to periodic to be consistent with the fields.");
```
Comment (dpgrote) [inline Source/Utils/WarpXUtil.cpp]: I don't think that this warning is needed if this behavior is documented.

**Comment by ax3l** [comment]
> Thanks for the changes! Can you add documentation for the new boundary input parameters?

Good idea, let's add it with this PR so code & doc stay in sync :)

**Comment by RevathiJambunathan** [comment]
Will do. Thank you for reviewing and for the suggestion @dpgrote and @ax3l . 
Would it be okay that the interface would only support periodic with this PR.
But PEC, PMC are coming up very soon!!!

File: Source/Utils/WarpXUtil.cpp (hunk 2 above)
Comment (RevathiJambunathan) [reply Source/Utils/WarpXUtil.cpp]: Thank you @dpgrote . I added documentation and will remove this Warning.

File: Docs/source/running_cpp/parameters.rst [hunk 3]
```diff
@@ -180,6 +180,29 @@ Setting up the field mesh
 * ``warpx.n_rz_azimuthal_modes`` (`integer`; 1 by default)
     When using the RZ version, this is the number of azimuthal modes.
 
+.. _running-cpp-parameters-parallelization:
+Domain Boundary Conditions
```
Comment (RevathiJambunathan) [inline Docs/source/running_cpp/parameters.rst:184]: @dpgrote and @ax3l May be I should add the documentation only for the BC that is supported in this PR, i.e., periodic?
So that we dont confuse users with options that are not currently supported through the interface? What do you think?
I have added it below anyway, but, as I was writing it out, I realized it may lead to confusion.

May be as the PRs come in to address these options, we can add documentation for that particle option?

Comment (dpgrote) [reply Docs/source/running_cpp/parameters.rst:184]: Yes, that would be good, to only add documentation for what is supported now. The rest can be added later with the subsequent PRs.

Comment (RevathiJambunathan) [reply Docs/source/running_cpp/parameters.rst:184]: sounds good. I will edit the doc above, and push the changes

File: Docs/source/running_cpp/parameters.rst [hunk 4]
```diff
@@ -180,6 +180,19 @@ Setting up the field mesh
...
+--------------------------
+
+* ``boundary.field_lo`` and ``boundary_field_hi`` (`2 strings` for 2D, `3 strings` for 3D)
+    Boundary conditions applied to field at the lower and upper domain boundaries.
+    Options are:
+    * ``Periodic``: This option can be used to set periodic domain boundaries. Note that if the fields for lo in a certain dimension are set to periodic, then the corresponding upper boundary must also be set to periodic. If particle boundaries are not specified in the input field, then particles boundaries by default will be set to periodic. If particles boundaries are specified, then they must be set to periodic corresponding to the periodic field boundaries.
```
Comment (dpgrote) [inline Docs/source/running_cpp/parameters.rst]: Is that a typo? "particle boundaries are not specified in the input field" should be "particle boundaries are not specified in the input file"?

Comment (RevathiJambunathan) [reply Docs/source/running_cpp/parameters.rst]: oh yes. thank yo for catching.

**Review by dpgrote — APPROVED** [review]
Looks good!

File: Source/Utils/WarpXAlgorithmSelection.H [hunk 5]
```diff
@@ -91,7 +91,35 @@ struct LoadBalanceCostsUpdateAlgo {
...
+        PMC = 3      //!< perfect magnetic conductor (PMC) with B_tangential=0
+    };
+};
+
+/** Particle boundary conditions at the domain boundary
+ */
+struct ParticleBoundaryType {
+    enum {
+        Absorbing = 0,     //!< particles crossing domain boundary are removed
+        Open = 1,          //!< particles cross domain boundary leave with damped j
```
Comment (atmyers) [inline Source/Utils/WarpXAlgorithmSelection.H:110]: Are Open and Reflecting implemented yet?

Comment (dpgrote) [reply Source/Utils/WarpXAlgorithmSelection.H:110]: Not yet. I have a separate PR that will do this, #1728

Comment (RevathiJambunathan) [reply Source/Utils/WarpXAlgorithmSelection.H:110]: oh this is a good point. Thank you @atmyers 
absorbing is implemented in the code, but, it is not interfaced with the input parameters in this PR and therefore not mentioned in the documentation.
But currently we have` particles.boundary_conditions = absorbing`, where particles are discarded when they leave the domain.
I think open is the same as `particle_j_damping` for particles entering the pml.
These will be interfaced with the input in subsequent PRs.

Comment (atmyers) [reply Source/Utils/WarpXAlgorithmSelection.H:110]: Ah, thanks!

**Review by atmyers — APPROVED** [review]

**Review by ax3l — APPROVED** [review]
